cpp/include/xtt-cpp/server_handshake_context.hpp

```diff
@@ -0,0 +1,110 @@
+#include <xtt-cpp/group_public_key_context.hpp>
+#include <xtt-cpp/types.hpp>
+#include <xtt-cpp/server_certificate_context.hpp>
+#include <xtt-cpp/server_cookie_context.hpp>
+
+#include <memory>
+#include <experimental/optional>
+
+namespace xtt { class server_handshake_context; }
+void swap(xtt::server_handshake_context&, xtt::server_handshake_context&);
+
+namespace xtt {
+
+class server_handshake_context {
+public:
+    server_handshake_context(const server_handshake_context&) = delete;
+
+    server_handshake_context& operator=(server_handshake_context other);
+
+    server_handshake_context(server_handshake_context&& other);
+
+    server_handshake_context(unsigned char *in_buffer,
+                             uint16_t in_buffer_size,
+                             unsigned char *out_buffer,
+                             uint16_t out_buffer_size);
+
+    std::experimental::optional<version> get_version() const;
+
+    std::experimental::optional<suite_spec> get_suite_spec() const;
+
+    std::unique_ptr<pseudonym> get_clients_pseudonym() const;
+
+    std::unique_ptr<longterm_key> get_clients_longterm_key() const;
+
+    std::experimental::optional<identity> get_clients_identity() const;
+
+    const struct xtt_server_handshake_context* get() const;
+    struct xtt_server_handshake_context* get();
+
+    return_code handle_io(uint16_t bytes_written,
+                          uint16_t bytes_read,
+                          uint16_t& io_bytes_requested,
+                          unsigned char*& io_ptr);
+
+    return_code handle_connect(uint16_t& io_bytes_requested,
+                               unsigned char*& io_ptr);
+
+    return_code build_serverattest(uint16_t& io_bytes_requested,
+                                   unsigned char*& io_ptr,
+                                   const server_certificate_context& certificate_ctx,
+                                   server_cookie_context& cookie_ctx);
+
+    return_code preparse_idclientattest(uint16_t& io_bytes_requested,
+                                        unsigned char*& io_ptr,
+                                        identity& requested_client_id_out,
+                                        group_identity& claimed_group_id_out,
+                                        server_cookie_context& cookie_ctx,
+                                        const server_certificate_context& certificate_ctx);
+
+    return_code verify_groupsignature(uint16_t& io_bytes_requested,
+                                      unsigned char*& io_ptr,
+                                      group_public_key_context& group_pub_key_ctx,
+                                      const server_certificate_context& certificate_ctx);
+
+    return_code build_idserverfinished(uint16_t& io_bytes_requested,
+                                       unsigned char*& io_ptr,
+                                       const identity& client_id);
+
+    return_code build_error_msg(uint16_t& io_bytes_requested,
```

Comment:
This is an `xtt` thing (not `xtt-cpp`) but the error message stuff really needs to be built in to `xtt`.  The user needs to be able to pass an error `xtt` lib, it prepares the error message in the out buffer, and it returns `WANT_WRITE` to have the message sent.  That's critical code that every user or wrapper maintainer has to reinvent and maintain right now.

Reply:
This comment may have been from an earlier version of the code (this stuff changed yesterday), but I think I agree with you and if so the changes made yesterday implement this.

Currently, `xtt` provides:
```
xtt_return_code_type
xtt_server_build_error_msg(uint16_t *io_bytes_requested,
                           unsigned char **io_ptr,
                           struct xtt_server_handshake_context* handshake_ctx);
```
and
```
xtt_return_code_type
xtt_client_build_error_msg(uint16_t *io_bytes_requested,
                           unsigned char **io_ptr,
                           struct xtt_client_handshake_context* handshake_ctx);
```
These two will build the error message in their internal buffers, set the `io_...` pointers, and return `WANT_WRITE`.

Is this what you're talking about?